openTSNE/affinity.py

```diff
@@ -237,9 +253,8 @@ def to_new(self, data, perplexity=None, return_distances=False):
         """
         perplexity = perplexity if perplexity is not None else self.perplexity
         perplexity = self.check_perplexity(perplexity)
-        k_neighbors = min(self.n_samples - 1, int(3 * perplexity))
 
-        neighbors, distances = self.knn_index.query(data, k_neighbors)
+        neighbors, distances = self.knn_index.query(data, self.k_neighbors)
```

Comment:
I am not a fan of this at all. For instance, if I create an embedding using 100k points using perplexity 500, then I want to add 1mln data points into that embedding, this means that we'll be calculating 500 neighbors for each 1mln data points. The number of neighbors can be pretty low in the transform functionality for it to work well, so this would really slow down `.transform` in that use case at very little benefit.

I think `transform` should have its own set of parameters, both for perplexity and k_neighbors.

Reply:
Fair enough. Fixed so that `3*perplexity` is used in `to_new()` unless this is larger than `self._k_neighbors`. If so, `self._k_neighbors` is used.

Reply:
Hmm, I'm not sure I see the point of this. Why would we want to limit the number of nearest neighbors when embedding  new data points to the number of nearest neighbors we used when constructing the embedding? They seem like two independent things to me. Any KNN index should be able to be queried with an arbitrary number of neighbors. If I wanted to, I might want to build an embedding using perplexity=30, then use perplexity=500 when adding new data points. I don't see any reason we might want to disallow this.

I think the number of neighbors should be completely seprate between the initial `fit` and `transform` functions.

Reply:
Okay - but you don't allow this in `set_perplexity()`. You want to allow increasing the size of the kNN graph in `to_new()` but not in `set_perplexity()`?

Reply:
Yes, that's right. It doesn't make that much sense, I know, but I added this when playing around with your perplexity annealing idea. I didn't see a reason to allow increasing the perplexity at the time (and I still don't, really). I don't think this is a function that gets used very much anyways.

Reply:
On reflection, I see the reason: in `set_perplexity()` we don't need to run `query()` but in `to_new` we do. I'll fix.

Reply:
Fixed.